File: growthscout-python/growthscout/streaming.py
# growthscout-python/growthscout/streaming.py
"""
SSE Event Stream consumer for Python SDK.
"""

import asyncio
import json
import logging
import httpx
from typing import AsyncGenerator, Optional
from datetime import datetime

from growthscout.exceptions import GrowthScoutNetworkError, GrowthScoutError
from growthscout.models import EventEnvelope

logger = logging.getLogger("growthscout.streaming")


class EventStream:
    """
    Parses incoming HTTP text/event-stream chunks into typed EventEnvelope objects.
    Supports reconnection logic with exponential backoff using the Last-Event-ID header.
    """
    def __init__(self, client, session_id: str) -> None:
        self.client = client
        self.session_id = session_id
        self.last_event_id: Optional[int] = None
        self.active = True

    async def listen(self) -> AsyncGenerator[EventEnvelope, None]:
        url = f"{self.client.base_url}/api/v1/sessions/{self.session_id}/stream"
        attempt = 0

        while self.active:
            headers = self.client._get_headers()
            if self.last_event_id is not None:
                headers["Last-Event-ID"] = str(self.last_event_id)

            try:
                async with httpx.AsyncClient(timeout=None) as client:
                    async with client.stream("GET", url, headers=headers) as response:
                        if response.status_code != 200:
                            self.client._handle_status_error(response.status_code, response.text)

                        # Successfully connected, reset retry attempt count
                        attempt = 0
                        buffer = ""
                        async for chunk in response.aiter_text():
                            if not self.active:
                                break
                            buffer += chunk
                            while "\n\n" in buffer:
                                block, buffer = buffer.split("\n\n", 1)
                                envelope = self._parse_block(block)
                                if envelope:
                                    self.last_event_id = envelope.event_id
                                    yield envelope
                                    if envelope.event_type == "stream_ended":
                                        self.active = False
                                        return

            except httpx.HTTPError as e:
                if not self.active:
                    break
                attempt += 1
                if attempt > self.client.max_retries:
                    raise GrowthScoutNetworkError(f"Reconnection failed after {attempt} attempts: {e}")

                sleep_time = min(10.0, 1.0 * (2 ** attempt))
                logger.warning(f"Connection dropped. Retrying in {sleep_time}s (attempt {attempt})...")
                await asyncio.sleep(sleep_time)

            except Exception as e:
                raise GrowthScoutError(f"Streaming error occurred: {e}")

    def close(self) -> None:
        """Closes the stream listener."""
        self.active = False

    def _parse_block(self, block: str) -> Optional[EventEnvelope]:
        lines = block.strip().split("\n")
        event_id = None
        event_type = None
        data = None

        for line in lines:
            if line.startswith("id:"):
                try:
                    event_id = int(line[3:].strip())
                except ValueError:
                    pass
            elif line.startswith("event:"):
                event_type = line[6:].strip()
            elif line.startswith("data:"):
                try:
                    data = json.loads(line[5:].strip())
                except json.JSONDecodeError:
                    pass

        if event_id is not None and event_type and data is not None:
            # Reconstruct envelope model
            # Skip heartbeats or return them as raw envelopes
            return EventEnvelope(
                event_id=event_id,
                event_type=event_type,
                session_id=data.get("session_id", self.session_id),
                timestamp=datetime.fromisoformat(data.get("timestamp", datetime.utcnow().isoformat())),
                data=data.get("data", {}),
                correlation_id=data.get("correlation_id")
            )
        return None

File: growthscout-js/src/streaming.ts
// growthscout-js/src/streaming.ts
/**
 * SSE Event Stream consumer for JavaScript/TypeScript SDK.
 */

import { EventEnvelope } from "./models.js";
import { GrowthScoutError, GrowthScoutNetworkError } from "./exceptions.js";

export class EventStream {
  private client: any;
  private sessionId: string;
  private lastEventId: number | null = null;
  private active: boolean = true;
  private listeners: Record<string, ((data: any) => void)[]> = {};

  constructor(client: any, sessionId: string) {
    this.client = client;
    this.sessionId = sessionId;
  }

  /**
   * Registers a callback listener for specific event types.
   */
  public on(event: string, callback: (data: any) => void): void {
    if (!this.listeners[event]) {
      this.listeners[event] = [];
    }
    this.listeners[event].push(callback);
  }

  /**
   * Closes the active stream connection.
   */
  public close(): void {
    this.active = false;
  }

  /**
   * Starts listening to the stream, handling chunks and reconnection automatically.
   */
  public async listen(): Promise<void> {
    const url = `${this.client.baseUrl}/api/v1/sessions/${this.sessionId}/stream`;
    let attempt = 0;

    while (this.active) {
      try {
        const headers = this.client.getHeaders();
        if (this.lastEventId !== null) {
          headers["Last-Event-ID"] = String(this.lastEventId);
        }

        const response = await fetch(url, {
          method: "GET",
          headers: headers,
        });

        if (!response.ok) {
          await this.client.handleStatusError(response.status, await response.text());
        }

        attempt = 0;
        const reader = response.body?.getReader();
        if (!reader) {
          throw new GrowthScoutError("Failed to obtain response stream reader.");
        }

        const decoder = new TextDecoder();
        let buffer = "";

        while (this.active) {
          const { done, value } = await reader.read();
          if (done || !this.active) {
            break;
          }

          buffer += decoder.decode(value, { stream: true });
          while (buffer.includes("\n\n")) {
            const index = buffer.indexOf("\n\n");
            const block = buffer.slice(0, index);
            buffer = buffer.slice(index + 2);

            const envelope = this.parseBlock(block);
            if (envelope) {
              this.lastEventId = envelope.event_id;
              this.emit(envelope.event_type, envelope);
              this.emit("message", envelope);

              if (envelope.event_type === "stream_ended") {
                this.active = false;
                return;
              }
            }
          }
        }
      } catch (err: any) {
        if (!this.active) break;
        attempt++;
        if (attempt > this.client.maxRetries) {
          const networkErr = new GrowthScoutNetworkError(`Reconnection failed after ${attempt} attempts: ${err.message}`);
          this.emit("error", networkErr);
          throw networkErr;
        }

        const sleepTime = Math.min(10000, 1000 * Math.pow(2, attempt));
        this.emit("warning", `Connection dropped. Retrying in ${sleepTime / 1000}s...`);
        await new Promise((resolve) => setTimeout(resolve, sleepTime));
      }
    }
  }

  private emit(event: string, data: any): void {
    const list = this.listeners[event] || [];
    for (const callback of list) {
      try {
        callback(data);
      } catch (err) {
        console.error("Error in stream callback", err);
      }
    }
  }

  private parseBlock(block: string): EventEnvelope | null {
    const lines = block.trim().split("\n");
    let eventId: number | null = null;
    let eventType: string | null = null;
    let data: any = null;

    for (const line of lines) {
      if (line.startsWith("id:")) {
        const val = parseInt(line.slice(3).trim(), 10);
        if (!isNaN(val)) eventId = val;
      } else if (line.startsWith("event:")) {
        eventType = line.slice(6).trim();
      } else if (line.startsWith("data:")) {
        try {
          data = JSON.parse(line.slice(5).trim());
        } catch {
          // ignore
        }
      }
    }

    if (eventId !== null && eventType && data !== null) {
      return {
        event_id: eventId,
        event_type: eventType,
        session_id: data.session_id || this.sessionId,
        timestamp: data.timestamp || new Date().toISOString(),
        data: data.data || {},
        correlation_id: data.correlation_id,
      };
    }
    return null;
  }
}
